pkg/util/inplaceupdate/inplace_update_defaults.go

```diff
@@ -100,55 +230,91 @@ func defaultCalculateInPlaceUpdateSpec(oldRevision, newRevision *apps.Controller
 	}
 
 	updateSpec := &UpdateSpec{
-		Revision:        newRevision.Name,
-		ContainerImages: make(map[string]string),
-		GraceSeconds:    opts.GracePeriodSeconds,
+		Revision:             newRevision.Name,
+		ContainerImages:      make(map[string]string),
+		ContainerRefMetadata: make(map[string]metav1.ObjectMeta),
+		GraceSeconds:         opts.GracePeriodSeconds,
 	}
 	if opts.GetRevision != nil {
 		updateSpec.Revision = opts.GetRevision(newRevision)
 	}
 
-	// all patches for podSpec can just update images
-	var metadataChanged bool
-	for _, jsonPatchOperation := range patches {
-		jsonPatchOperation.Path = strings.Replace(jsonPatchOperation.Path, "/spec/template", "", 1)
+	// all patches for podSpec can just update images in pod spec
+	var metadataPatches []jsonpatch.Operation
+	for _, op := range patches {
+		op.Path = strings.Replace(op.Path, "/spec/template", "", 1)
 
-		if !strings.HasPrefix(jsonPatchOperation.Path, "/spec/") {
-			metadataChanged = true
-			continue
+		if !strings.HasPrefix(op.Path, "/spec/") {
+			if strings.HasPrefix(op.Path, "/metadata/") {
+				metadataPatches = append(metadataPatches, op)
+				continue
+			}
+			return nil
 		}
-		if jsonPatchOperation.Operation != "replace" || !inPlaceUpdatePatchRexp.MatchString(jsonPatchOperation.Path) {
+		if op.Operation != "replace" || !containerImagePatchRexp.MatchString(op.Path) {
 			return nil
 		}
 		// for example: /spec/containers/0/image
-		words := strings.Split(jsonPatchOperation.Path, "/")
+		words := strings.Split(op.Path, "/")
 		idx, _ := strconv.Atoi(words[3])
 		if len(oldTemp.Spec.Containers) <= idx {
 			return nil
 		}
-		updateSpec.ContainerImages[oldTemp.Spec.Containers[idx].Name] = jsonPatchOperation.Value.(string)
+		updateSpec.ContainerImages[oldTemp.Spec.Containers[idx].Name] = op.Value.(string)
 	}
-	if metadataChanged {
+
+	if len(metadataPatches) > 0 {
```

Comment:
no need to check len, iteration through empty map will do nothing

Reply:
There is not only the iteration, but also calculate strategic patch using old and new templates.